quesma/jsondiff/jsondiff_test.go

```diff
@@ -0,0 +1,127 @@
+// Copyright Quesma, licensed under the Elastic License 2.0.
+// SPDX-License-Identifier: Elastic-2.0
+package jsondiff
+
+import (
+	"fmt"
+	"quesma/quesma/types"
+	"testing"
+)
+
+func TestJSONDiff(t *testing.T) {
+
+	problem := func(path string, problemType problemType) Problem {
+		return Problem{
+			Path: path,
+			Type: problemType.code,
+		}
+	}
+
+	tests := []struct {
+		name     string
+		expected string
+		actual   string
+		problems []Problem
+	}{
+		{
+			name:     "Test 1",
+			expected: `{"a": 1, "b": 2, "c": 3}`,
+			actual:   `{"a": 1, "b": 2, "c": 3}`,
+			problems: []Problem{},
+		},
+
+		{
+			name:     "Test 2",
+			expected: `{"a": 1, "b": 2, "c": 3}`,
+			actual:   `{"a": 1, "b": 3, "c": 3}`,
+			problems: []Problem{problem("b", invalidValue)},
+		},
+
+		{
+			name:     "invalid type",
+			expected: `{"a": 1, "b": 2, "c": 3}`,
+			actual:   `{"a": 1, "b": "foo", "c": 3}`,
+			problems: []Problem{problem("b", invalidType)},
+		},
+
+		{
+			name:     "missing value",
+			expected: `{"a": 1, "b": 2, "c": 3}`,
+			actual:   `{"a": 1, "c": 3}`,
+			problems: []Problem{problem("b", invalidValue)},
+		},
+
+		{
+			name:     "array length",
+			expected: `{"a": [1, 2, 3], "b": 2, "c": 3}`,
+			actual:   `{"a": [1, 2], "b": 2, "c": 3}`,
+			problems: []Problem{problem("a", invalidArrayLength)},
+		},
+
+		{
+			name:     "array element difference",
+			expected: `{"a": [1, 2, 3], "b": 2, "c": 3}`,
+			actual:   `{"a": [1, 2, 4], "b": 2, "c": 3}`,
+			problems: []Problem{problem("a.[2]", invalidValue)},
+		},
+
+		{
+			name:     "array element difference",
+			expected: `{"a": [1, 2, 3]}`,
+			actual:   `{"a": [1, true, "xx"]}`,
+			problems: []Problem{problem("a.[1]", invalidType), problem("a.[2]", invalidType)},
+		},
+
+		{
+			name:     "object difference",
+			expected: `{"a": {"b": 1}, "c": 3}`,
+			actual:   `{"a": {"b": 2}, "c": 3}`,
+			problems: []Problem{problem("a.b", invalidValue)},
+		},
+
+		{
+			name:     "deep path difference",
+			expected: `{"a": {"d": {"b": 1}}, "c": 3}`,
+			actual:   `{"a": {"d": {"b": 2}}, "c": 3}`,
+			problems: []Problem{problem("a.d.b", invalidValue)},
+		},
+
+		{
+			name:     "deep path difference",
+			expected: `{"a": {"d": {"b": 1}}, "c": 3, "_ignore": 1}`,
```

Comment:
What with the case where one json has more keys than the other? Should it be reported?

Reply:
Not all corner cases are covered in this PR.  I will fix it in the next ones.